### PR TITLE
Fix R_AVR_7_PCREL and R_AVR_13_PCREL range checking

### PR DESCRIPTION
**Pull request by aykevl**
See the commit messages for detail. The first one fixes a bug I found while updating to LLVM 18 (see https://github.com/llvm/llvm-project/pull/67636#issuecomment-2118869227), the second one fixes a bug I found while fixing the first.

I'm not sure whether sending two commits in a single PR like this is allowed, but since they're related (but fix different bugs) it made sense to me to do it in one PR.

### CODE REVIEW

File: lld/test/ELF/avr-reloc.s [hunk 1]
```diff
@@ -82,13 +82,25 @@ sbic  b, 1    ; R_AVR_PORT5
...
 rjmp foo + 32  ; R_AVR_13_PCREL
 rjmp foo - 32  ; R_AVR_13_PCREL
 breq foo + 32  ; R_AVR_7_PCREL
 breq foo - 32  ; R_AVR_7_PCREL
+rjmp 1f - 4096  $ 1:  ; R_AVR_13_PCREL
+rjmp 1f + 4094  $ 1:  ; R_AVR_13_PCREL
+rjmp 1f - 4098  $ 1:  ; R_AVR_13_PCREL (overflow)
+rjmp 1f + 4096  $ 1:  ; R_AVR_13_PCREL (overflow)
+breq 1f - 128   $ 1:  ; R_AVR_7_PCREL
+breq 1f + 126   $ 1:  ; R_AVR_7_PCREL
```
Comment (aykevl) [inline lld/test/ELF/avr-reloc.s:103]: For those not familiar with this syntax, the `$` symbol works as a newline.
So this code:
```s
rjmp 1f - 4096  $ 1:
```

Is equivalent to the following:

```s
rjmp 1f - 4096
1:
```
(where `1f` refers to the next anonymous label numbered "1")

**Review by MaskRay — APPROVED** [review]

**Comment by benshi001** [comment]
I agree with the change to `R_AVR_PCREL_7`, there are 7 bits in the `RJMP`/`RCALL` instruction encoding which actually represents [-256, 254].

**Comment by aykevl** [comment]
Thanks for the review!
@benshi001 do you also agree with the `R_AVR_13_PCREL` change?

**Comment by benshi001** [comment]
I wonder how RJMP can jumps over 4kB on the ATtiny85 device. Is RJMP has a different encoding against other devices ?

**Comment by aykevl** [comment]
> I wonder how RJMP can jumps over 4kB on the ATtiny85 device. Is RJMP has a different encoding against other devices ?

I can't quickly find it in the instruction set manual, but I did find this:

> For AVR microcontrollers with Program memory not exceeding 4K words (8KB) this instruction can address the entire memory from every address location.

I assume it wraps around, because that's the only way I see that would make it possible to address the full 8kB of memory.
So for example if there's a `rjmp .+4094` at address 6114, it will jump to (6114 + 4094 + 2) = 10210, which wraps around to address 2018. (The +2 is the size of the instruction itself).

I also found this in the [attiny85 datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-2586-AVR-8-bit-Microcontroller-ATtiny25-ATtiny45-ATtiny85_Datasheet.pdf):

> The ATtiny25/45/85 Program Counter (PC) is 10/11/12 bits wide, thus addressing the 1024/2048/4096 Program memory locations

This means that the attiny85 program counter is 12 bits, which means that it will indeed wrap around (12 bits is 4096 words, is 8192 bytes).

For avr-ld, there's an option to enable/disable wraparound (but it appears that wraparound is enabled by default):
https://sourceware.org/binutils/docs-2.30/as/AVR-Options.html. See `-mno-wrap`. I don't know whether we can add a flag like that to lld, I'd much prefer if we didn't need a special chip-specific flag for this (in my opinion a better solution would be two different relocations: one with wraparound and one without - but that would break compatibility with avr-ld).

**Comment by benshi001** [comment]
> > I wonder how RJMP can jumps over 4kB on the ATtiny85 device. Is RJMP has a different encoding against other devices ?
> 
> I can't quickly find it in the instruction set manual, but I did find this:
> 
> > For AVR microcontrollers with Program memory not exceeding 4K words (8KB) this instruction can address the entire memory from every address location.
> 
> I assume it wraps around, because that's the only way I see that would make it possible to address the full 8kB of memory. So for example if there's a `rjmp .+4094` at address 6114, it will jump to (6114 + 4094 + 2) = 10210, which wraps around to address 2018. (The +2 is the size of the instruction itself).
> 
> I also found this in the [attiny85 datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-2586-AVR-8-bit-Microcontroller-ATtiny25-ATtiny45-ATtiny85_Datasheet.pdf):
> 
> > The ATtiny25/45/85 Program Counter (PC) is 10/11/12 bits wide, thus addressing the 1024/2048/4096 Program memory locations
> 
> This means that the attiny85 program counter is 12 bits, which means that it will indeed wrap around (12 bits is 4096 words, is 8192 bytes).
> 
> For avr-ld, there's an option to enable/disable wraparound (but it appears that wraparound is enabled by default): https://sourceware.org/binutils/docs-2.30/as/AVR-Options.html. See `-mno-wrap`. I don't know whether we can add a flag like that to lld, I'd much prefer if we didn't need a special chip-specific flag for this (in my opinion a better solution would be two different relocations: one with wraparound and one without - but that would break compatibility with avr-ld).

I would agree that lld should support wraparound  as default, but could you please supplement some encoding tests for that in [lld/test/ELF/avr-reloc.s](https://github.com/llvm/llvm-project/pull/92695#diff-020c5c5113ee98bd23be19a24d0347a8cf5d149c4758a9ac5379861206637696), just like the normal encoding tests.

**Comment by benshi001** [comment]
> > I wonder how RJMP can jumps over 4kB on the ATtiny85 device. Is RJMP has a different encoding against other devices ?
> 
> I can't quickly find it in the instruction set manual, but I did find this:
> 
> > For AVR microcontrollers with Program memory not exceeding 4K words (8KB) this instruction can address the entire memory from every address location.
> 
> I assume it wraps around, because that's the only way I see that would make it possible to address the full 8kB of memory. So for example if there's a `rjmp .+4094` at address 6114, it will jump to (6114 + 4094 + 2) = 10210, which wraps around to address 2018. (The +2 is the size of the instruction itself).
> 
> I also found this in the [attiny85 datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-2586-AVR-8-bit-Microcontroller-ATtiny25-ATtiny45-ATtiny85_Datasheet.pdf):
> 
> > The ATtiny25/45/85 Program Counter (PC) is 10/11/12 bits wide, thus addressing the 1024/2048/4096 Program memory locations
> 
> This means that the attiny85 program counter is 12 bits, which means that it will indeed wrap around (12 bits is 4096 words, is 8192 bytes).
> 
> For avr-ld, there's an option to enable/disable wraparound (but it appears that wraparound is enabled by default): https://sourceware.org/binutils/docs-2.30/as/AVR-Options.html. See `-mno-wrap`. I don't know whether we can add a flag like that to lld, I'd much prefer if we didn't need a special chip-specific flag for this (in my opinion a better solution would be two different relocations: one with wraparound and one without - but that would break compatibility with avr-ld).

LGTM!

**Review by benshi001 — APPROVED** [review]

File: lld/ELF/Arch/AVR.cpp [hunk 2]
```diff
@@ -238,7 +238,6 @@ void AVR::relocate(uint8_t *loc, const Relocation &rel, uint64_t val) const {
     break;
   }
   case R_AVR_13_PCREL: {
-    checkInt(loc, val - 2, 13, rel);
```
[the remark this answers is not in the document]
Comment (aykevl) [reply lld/ELF/Arch/AVR.cpp:241]: Looks like these were older comments?

**Comment by aykevl** [comment]
Thank you both for the review!

Can someone merge this PR? (I don't have commit access anymore). @MaskRay?

**Pull request merged**